[bitnami/kube-prometheus] add thanos ruler CR

kir4h opened the pull request:
<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

### Description of the change

This PR adds support for creating a [ThanosRuler CR](https://prometheus-operator.dev/docs/getting-started/design/#thanosruler).

`ThanosRuler` is already available as a `Statefulset` on [Bitnami Thanos chart](https://github.com/kir4h/bitnami-charts/tree/main/bitnami/thanos/templates/ruler), but there is no option to create it using the CRD deployed by Prometheus Operator. After some brainstorming on https://github.com/bitnami/charts/issues/32826 , the decision was to [add it as part of kube-prometheus chart](https://github.com/bitnami/charts/issues/32826#issuecomment-2817796931).

cc @juan131 

### Benefits

Using a CR for Thanos ruler takes away complexity from the Statefulset, but also allows the usage of [PrometheusRule](https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.PrometheusRule) objects to define rules in a more flexible manner (using rule selectors to state which Thanos Ruler is the target for a given rule, or allowing to decouple the deployment of rules from ruler itself).

This is already the way [kube-prometheus-stack deploys it](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/templates/thanos-ruler/ruler.yaml).

### Possible drawbacks

None that I can think of.

### Applicable issues

- fixes #32826 
- fixes #28525


### Additional information

<!-- If there's anything else that's important and relevant to your pull request, mention that information here.-->

I have done some rendering tests with different configurations and a basic deployment test with:

* [Thanos chart](https://github.com/bitnami/charts/tree/main/bitnami/thanos) with default values
* This chart with default values, but enabling ruler

```console
❯ helm ls
NAME           	NAMESPACE	REVISION	UPDATED                              	STATUS  	CHART                 	APP VERSION
kube-prometheus	default  	3       	2025-05-20 09:46:41.319475 +0200 CEST	deployed	kube-prometheus-11.2.0	0.82.2
thanos         	default  	3       	2025-05-13 08:26:18.067957 +0200 CEST	deployed	thanos-16.0.6         	0.38.0
```

Values used for kube-prometheus

```yaml
thanosRuler:
  enabled: true
  queryConfig:
    config:
    - static_configs:
      - "dnssrv+_http._tcp.thanos-query.default.svc.cluster.local"
```

Ruler running

```console
❯ k get po | grep ruler
thanos-ruler-kube-prometheus-thanos-ruler-0           2/2     Running   4 (23h ago)   7d
```

Ruler UI looks as expected

<img width="1503" alt="image" src="https://github.com/user-attachments/assets/68102b15-afc2-40f8-86e1-7829a4a7ebe0" />

<img width="2523" alt="image" src="https://github.com/user-attachments/assets/4cf423a6-8ec9-4644-8286-09cc69ef327d" />



### Checklist

<!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

juan131 commented on bitnami/kube-prometheus/values.yaml:
```diff
@@ -3307,3 +3307,565 @@ kubeProxy:
+  containerPorts:
+    http: 10902
+    grpc: 10901
+  ## @param thanosRuler.alertQueryUrl The external Query URL the Thanos Ruler will set in the ‘Source’ field of all alerts
+  ## Maps to the ‘–alert.query-url’ CLI arg
+  ##
+  alertQueryUrl: ""
+  ## Configuration for connecting to alertmanager
+  ## Maps to --alertmanagers.config Thanos Ruler argument
+  ## NOTE: This field takes precedence over alertmanagersUrl.
+  ##
+  alertmanagersConfig:
+    ## @param thanosRuler.alertmanagersConfig.existingSecret.name Name of an existing secret to use for Alert Manager config
+    ## If configured, thanosRuler.alertmanagersConfig.config will not be used
+    ## @param thanosRuler.alertmanagersConfig.existingSecret.key Name of a key in the existing secret to use for Alert Manager config
+    ##
+    existingSecret:
+      name: ""
+      key: ""
+    ## @param thanosRuler.alertmanagersConfig.config
+    ## Unused if thanosRuler.alertmanagersConfig.existingSecret.name is configured
+    ## ref: https://thanos.io/tip/components/rule.md/#alertmanager
+    ## e.g:
+    ##   alertmanagers:
+    ##     - http_config:
+    ##         basic_auth:
+    ##           username: some_user
+    ##           password: some_pass
+    ##       static_configs:
+    ##         - alertmanager.thanos.io
+    ##       scheme: http
+    ##       timeout: 10s
+    ##       api_version: v2
+    ##
+    config: {}
```
Shouldn't this configuration be auto-generated base on `alertmanager.*` parameters?

kir4h replied:
Definitely. Again, ported from Thanos chart, that didn't include Alertmanager.

I'm thinking of having:
* ability to inject config through existing secret (already in place)
* a default config that uses `alertmanager.*` parameters to point to alertmanager deployed by this chart

What I'm not sure about:
* should we add ability not to have any alertmanager config even if the chart deploys one?
* should we add ability to deploy a static config ignoring the one deployed by the chart? 

Both are edge cases in my mind, typically if you deploy an alertmanager and a ruler with this chart, you will want ruler to talk to that Alertmanager. But maybe there are scenarios and adding support for that will give more flexibility?

juan131 commented on bitnami/kube-prometheus/values.yaml:
```diff
@@ -3307,3 +3307,565 @@ kubeProxy:
+  ## @param thanosRuler.priorityClassName Priority class assigned to the Pods
+  ##
+  priorityClassName: ""
+  ## @param thanosRuler.portName Port name used for the pods and governing service. Defaults to web
+  ##
+  portName: "web"
```
Default value not required in parameter description
```suggestion
  ## @param thanosRuler.portName Port name used for the pods and governing service
  ##
  portName: web
```

juan131 reviewed (changes requested):
Thanks so much for this amazing contribution, we really appreciate it! Please take a look to my comments when you have a chance.

kir4h commented on bitnami/kube-prometheus/Chart.yaml:
```diff
@@ -51,4 +51,4 @@ maintainers:
 sources:
 - https://github.com/bitnami/charts/tree/main/bitnami/kube-prometheus
-version: 11.1.13
+version: 11.2.0
```
> Ensure you add the new container image under `annotations.images`

Thanos uses a single binary, so also a single docker image. The image was already present at https://github.com/bitnami/charts/blob/main/bitnami/kube-prometheus/Chart.yaml#L15-L16 since it's used for Thanos sidecar. 

I am not sure of the purpose of the key under `annotation.images` ; I think as it's metadata it's used for chart info so that entry should cover every usage of the image (please let me wrong if otherwise and I need to add an extra entry with different key)

kir4h commented on bitnami/kube-prometheus/templates/thanos-ruler/networkpolicy.yaml:
```diff
@@ -0,0 +1,113 @@
+          protocol: UDP
+        - port: 53
+          protocol: TCP
+    # Communicate with other ruler instances via headless service
+    - ports:
+        - port: {{ .Values.thanosRuler.containerPorts.http }}
+        - port: {{ .Values.thanosRuler.containerPorts.grpc }}
+      to:
+        - podSelector:
+            matchLabels: {{ include "common.labels.matchLabels" ( dict "customLabels" $podLabels "context" $ ) | nindent 14 }}
+              app.kubernetes.io/component: thanos-ruler
```
Good point! I used Thanos chart as base, so I copied over that rule from https://github.com/bitnami/charts/blob/main/bitnami/thanos/templates/ruler/networkpolicy.yaml#L36-L43

But I see no reason why rulers should talk to each other, and haven't found any signs of it under Thanos docs. 
Since ruler is kind of a "piece of Prometheus", I checked Prometheus docs and on HA mode they don't talk to each other (https://promlabs.com/blog/2023/08/31/high-availability-for-prometheus-and-alertmanager-an-overview/). 

Additionally, ruler configuration doesn't include any argument related to service discovery of other rulers, so I'm removing this one. (and following lines since they reference minio, which is not part of this chart)

kir4h commented on bitnami/kube-prometheus/values.yaml:
```diff
@@ -3307,3 +3307,565 @@ kubeProxy:
+    ## @param thanosRuler.serviceAccount.automountServiceAccountToken Automount service account token for the server service account
+    ##
+    automountServiceAccountToken: false
+  ## Configure Security Context
+  ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod
+  ## @param thanosRuler.securityContext Holds pod-level security attributes and common container settings. This defaults to the default PodSecurityContext
+  ##  Example:
+  ##  securityContext:
+  ##    runAsGroup: 1001
+  ##    runAsNonRoot: true
+  ##    runAsUser: 1001
+  ##    fsGroup: 1001
+  ##    seccompProfile:
+  ##    type: RuntimeDefault
+  securityContext: {}
+  ## Configure Container Security Context
+  ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-container
+  ## @param thanosRuler.containerSecurityContext.enabled Enable container security context
+  ## @param thanosRuler.containerSecurityContext.runAsUser User ID for the container
+  ## @param thanosRuler.containerSecurityContext.runAsNonRoot Enable running as non-root
+  ##
+  containerSecurityContext:
+    enabled: true
+    runAsUser: 1001
+    runAsNonRoot: true
```
Thanks! missed adapting this one (kept the format from the other chart). I will adapt it.

Regarding `containerSecurityContext`, I've been reading and my preferred approach would be not to include it.
Rationale:
* Not supported directly by the operator. We currently do some strategic merge patching to include it
* From https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.ThanosRulerSpec  
> The current container names are: thanos-ruler and config-reloader. Overriding containers is entirely outside the scope of what the maintainers will support and by doing so, you accept that this behaviour may break at any time without notice.
* Operator containers have some security restrictions already in place, which I think should cover most cases https://github.com/prometheus-operator/prometheus-operator/commit/d46ae5145ed96b69647741f79100326d2f2a4786

wdyt?

kir4h commented on bitnami/kube-prometheus/values.yaml:
```diff
@@ -3307,3 +3307,565 @@ kubeProxy:
+  ## ref: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
+  ##
+  resources: {}
+  ## @param thanosRuler.containerPorts.http HTTP container port
+  ## @param thanosRuler.containerPorts.grpc GRPC container port
+  ##
+  containerPorts:
+    http: 10902
+    grpc: 10901
```
good catch! I need to inject them as `additionalArgs`

kir4h commented:
> Thanks so much for this amazing contribution, we really appreciate it! Please take a look to my comments when you have a chance.

Thanks for the quick review! Tough one to review since it's a long PR.
 I have started implementing the requested changes (applied all suggestions, need to finish some of the comments) will continue tomorrow morning

kir4h commented:
@juan131 I've submitted the changes.

Some comments:

* disabled alertmanager configuration on prometheus when thanos ruler is enabled ( I think it makes sense, since we don't want both sending the same alerts)
* ports are not exposed through the CR, so had to use the `additionalArgs` field. I think I need to refine the readme generator annotation, I added a skip and now defaults to "undefined", not sure what I should force default to in description (I could also hardcode these port values in the templates, and then concatenate `additionalArgs` leaving it empty by default if you think it's better
* for alertmanager config, I have added a third scenario 1) use existing secret if provided 2) use config if provided 3) fallback to chart alertmanager if alertmanager enabled

Regarding conversation resolution on this PR: should I resolve my self or let you do it?

juan131 commented on bitnami/kube-prometheus/templates/prometheus/prometheus.yaml:
```diff
@@ -41,7 +41,7 @@ spec:
     {{- if .Values.prometheus.alertingEndpoints }}
     {{- include "common.tplvalues.render" (dict "value" .Values.prometheus.alertingEndpoints "context" $) | nindent 6 }}
-    {{- else if .Values.alertmanager.enabled }}
+    {{- else if and .Values.alertmanager.enabled (not .Values.thanosRuler.enabled) }}
```
Yes, it's okay to use `kube-prometheus.validateValues`

juan131 commented on bitnami/kube-prometheus/templates/thanos-ruler/ingress.yaml:
```diff
@@ -0,0 +1,59 @@
+  name: {{ include "kube-prometheus.thanosRuler.fullname" . }}
+  namespace: {{ include "common.names.namespace" . }}
+  {{- $versionLabel := dict "app.kubernetes.io/version" (include "common.images.version" (dict "imageRoot" .Values.thanosRuler.image "chart" .Chart)) }}
+  {{- $labels := include "common.tplvalues.merge" (dict "values" (list .Values.commonLabels $versionLabel) "context" .) }}
```
```suggestion
  {{- $labels := include "common.tplvalues.merge" (dict "values" (list .Values.thanosRuler.ingress.labels .Values.commonLabels $versionLabel) "context" .) }}
```

juan131 commented on bitnami/kube-prometheus/templates/thanos-ruler/secrets-tls.yaml:
```diff
@@ -0,0 +1,48 @@
+{{- /*
```
Please rename this file to `ingress-tls-secret.yaml`

juan131 commented on bitnami/kube-prometheus/templates/thanos-ruler/service.yaml:
```diff
@@ -0,0 +1,64 @@
+  name: {{ include "kube-prometheus.thanosRuler.fullname" . }}
+  namespace: {{ include "common.names.namespace" . }}
+  {{- $versionLabel := dict "app.kubernetes.io/version" (include "common.images.version" (dict "imageRoot" .Values.thanosRuler.image "chart" .Chart)) }}
+  {{- $labels := include "common.tplvalues.merge" (dict "values" (list .Values.commonLabels $versionLabel) "context" .) }}
```
```suggestion
  {{- $labels := include "common.tplvalues.merge" (dict "values" (list .Values.thanosRuler.service.labels .Values.commonLabels $versionLabel) "context" .) }}
```

juan131 commented on bitnami/kube-prometheus/templates/thanos-ruler/servicemonitor.yaml:
```diff
@@ -0,0 +1,48 @@
+  name: {{ include "kube-prometheus.thanosRuler.fullname" . }}
+  namespace: {{ include "common.names.namespace" . }}
+  {{- $versionLabel := dict "app.kubernetes.io/version" (include "common.images.version" (dict "imageRoot" .Values.thanosRuler.image "chart" .Chart)) }}
+  {{- $labels := include "common.tplvalues.merge" (dict "values" (list .Values.commonLabels $versionLabel) "context" .) }}
```
```suggestion
  {{- $labels := include "common.tplvalues.merge" (dict "values" (list .Values.thanosRuler.serviceMonitor.labels .Values.commonLabels $versionLabel) "context" .) }}
```

juan131 reviewed (commented):
This is almost ready! Just a few very minor comments

juan131 reviewed (approved):
Amazing contribution! LGTM
